File: logic/sim/jztb/src/ztb.ts
import {
  Trace as TraceModule, TraceTypes,
} from '@pepnext/device-interface';

type Trace<T> = TraceTypes.Trace<T>;
type TraceBuffer = TraceModule.TraceBuffer
type TraceBufferStatus = TraceModule.TraceBufferStatus
type TraceCommitHook = TraceModule.TraceCommitHook

// eslint-disable-next-line import/prefer-default-export
export class ZTB implements TraceBuffer {
  // eslint-disable-next-line class-methods-use-this,@typescript-eslint/no-empty-function
  discard(): IterableIterator<Trace<any>> {
    return [][Symbol.iterator]();
  }

  // eslint-disable-next-line class-methods-use-this,@typescript-eslint/no-empty-function
  pop(): void {

  }

  // eslint-disable-next-line class-methods-use-this,no-undef
  push(trace: Trace<any>|Array<Trace<any>>): TraceBufferStatus {
    // eslint-disable-next-line no-void
    void trace;
    return { success: true, overflow: false };
  }

  // eslint-disable-next-line class-methods-use-this,@typescript-eslint/no-empty-function
  stage(): void {

  }

  // eslint-disable-next-line class-methods-use-this
  traceDevice(device: number, enabled: boolean): void {
    // eslint-disable-next-line no-void
    void [device, enabled];
  }

  // eslint-disable-next-line class-methods-use-this
  pending(): IterableIterator<Trace<any>> {
    return [][Symbol.iterator]();
  }

  // eslint-disable-next-line class-methods-use-this
  staged(): IterableIterator<Trace<any>> {
    return [][Symbol.iterator]();
  }

  // eslint-disable-next-line class-methods-use-this,@typescript-eslint/no-unused-vars,no-empty-function,@typescript-eslint/no-empty-function
  commit(strategy: 'preferred' | 'all'): void {
  }

  // eslint-disable-next-line class-methods-use-this,@typescript-eslint/no-unused-vars
  registerCommitHook(hook: TraceCommitHook): number {
    return 0;
  }

  // eslint-disable-next-line class-methods-use-this,@typescript-eslint/no-unused-vars
  unregisterCommitHook(hookID: number): TraceCommitHook|undefined {
    return undefined;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars,class-methods-use-this,@typescript-eslint/no-empty-function
  tick(currentTick: number): void {
  }
}

File: logic/sim/cztb/src/sim/trace/ztb.cpp

#include "ztb.hpp"
#include "cabi/trace/descriptor.h"

CTraceBuffer sim::trace::ZTB::as_trace_buffer() {
  auto x = CTraceBuffer{
      .tick = ZTB::tick_static,
      .push=ZTB::push_static,
      .pop=ZTB::pop_static,
      .pending=ZTB::pending_static,
      .stage=ZTB::stage_static,
      .discard=ZTB::discard_static,
      .staged=ZTB::staged_static,
      .commit=ZTB::commit_static,
      .register_commit_hook=ZTB::register_commit_hook_static,
      .unregister_commit_hook=ZTB::unregister_commit_hook_static,
      .trace_device=ZTB::trace_device_static,
      .impl = this};
  return x;
}

void sim::trace::ZTB::tick_static(void *impl, uint64_t current_tick) {

}

TraceBufferStatus sim::trace::ZTB::push(TraceDescriptorChunk &traces) {
  for (auto it = 0; it < traces.length; it++) {
    auto trace = traces.traces[it];
    this->_free_traces(trace.type.as_uint16, trace.payload);
  }
  return TraceBufferStatus::Success;

}
TraceBufferStatus sim::trace::ZTB::push_static(void *impl, TraceDescriptorChunk &traces) {
  return static_cast<ZTB *>(impl)->push(traces);
}
void sim::trace::ZTB::pop() {

}

void sim::trace::ZTB::pop_static(void *impl) {

}
NonOwningTraceDescriptorChunkList *sim::trace::ZTB::pending_static(void *impl) {
  return new NonOwningTraceDescriptorChunkList{.chunk={.length=0, .traces=nullptr}};
}

void sim::trace::ZTB::stage_static(void *impl) {

}

OwningTraceDescriptorChunkList *sim::trace::ZTB::discard_static(void *impl) {
  return new OwningTraceDescriptorChunkList{.chunk={.length=0, .traces=nullptr}};
}

NonOwningTraceDescriptorChunkList *sim::trace::ZTB::staged_static(void *impl) {
  return new NonOwningTraceDescriptorChunkList{.chunk={.length=0, .traces=nullptr}};
}

void sim::trace::ZTB::commit_static(void *impl, TraceBufferCommitStrategy strategy) {

}

uint16_t sim::trace::ZTB::register_commit_hook_static(void *impl, void *hook_impl, commit_hook_fn hook_fn) {
  return 0;
}

bool sim::trace::ZTB::unregister_commit_hook_static(void *impl, uint16_t hook_id) {
  return false;
}

void sim::trace::ZTB::trace_device_static(void *this_ptr, device_id_t device, bool enabled) {

}



